src/gradient/gradient.class.ts

```diff
@@ -1,493 +1,334 @@
+import { ColorStop, GradientCoords, GradientOptions, GradientType, GradientUnits, SVGOptions } from "./typedefs";
+
+/**
+ * @todo remove this transient junk
+ */
+type FabricObject = any;
+
+/**
+ * Gradient class
+ * @class Gradient
+ * @tutorial {@link http://fabricjs.com/fabric-intro-part-2#gradients}
+ */
+export class Gradient<S, T extends GradientType = S extends GradientType ? S : 'linear'> {
 
-(function(global) {
-  var fabric = global.fabric;
-  /* _FROM_SVG_START_ */
-  function getColorStop(el, multiplier) {
-    var style = el.getAttribute('style'),
-        offset = el.getAttribute('offset') || 0,
-        color, colorAlpha, opacity, i;
+  /**
+   * Horizontal offset for aligning gradients coming from SVG when outside pathgroups
+   * @type Number
+   * @default 0
+   */
+  offsetX = 0
 
-    // convert percents to absolute values
-    offset = parseFloat(offset) / (/%$/.test(offset) ? 100 : 1);
-    offset = offset < 0 ? 0 : offset > 1 ? 1 : offset;
-    if (style) {
-      var keyValuePairs = style.split(/\s*;\s*/);
+  /**
+   * Vertical offset for aligning gradients coming from SVG when outside pathgroups
+   * @type Number
+   * @default 0
+   */
+  offsetY = 0
 
-      if (keyValuePairs[keyValuePairs.length - 1] === '') {
-        keyValuePairs.pop();
-      }
+  /**
+   * A transform matrix to apply to the gradient before painting.
+   * Imported from svg gradients, is not applied with the current transform in the center.
+   * Before this transform is applied, the origin point is at the top left corner of the object
+   * plus the addition of offsetY and offsetX.
+   * @type Number[]
+   * @default null
+   */
+  gradientTransform: number[] | null = null
 
-      for (i = keyValuePairs.length; i--; ) {
+  /**
+   * coordinates units for coords.
+   * If `pixels`, the number of coords are in the same unit of width / height.
+   * If set as `percentage` the coords are still a number, but 1 means 100% of width
+   * for the X and 100% of the height for the y. It can be bigger than 1 and negative.
+   * allowed values pixels or percentage.
+   * @type GradientUnits
+   * @default 'pixels'
+   */
+  gradientUnits: GradientUnits
 
-        var split = keyValuePairs[i].split(/\s*:\s*/),
-            key = split[0].trim(),
-            value = split[1].trim();
+  /**
+   * Gradient type linear or radial
+   * @type GradientType
+   * @default 'linear'
+   */
+  type: T
+
+  coords: GradientCoords<T>
+
+  colorStops: ColorStop[]
+
+  private id: string | number
+
+  constructor({
+    type = 'linear' as T,
+    gradientUnits = 'pixels',
+    coords,
+    colorStops = [],
+    offsetX = 0,
+    offsetY = 0,
+    gradientTransform,
+    id
+  }: GradientOptions<T>) {
+    const uid = fabric.Object.__uid++;
+    this.id = id ? `${id}_${uid}` : uid;
+    this.type = type;
+    this.gradientUnits = gradientUnits;
+    this.gradientTransform = gradientTransform || null;
+    this.offsetX = offsetX;
+    this.offsetY = offsetY;
```

Comment:
ok so is there any reason why we want to specify those manually while before we weren't?

Coords and colorStops make sense, but the rest i would rather assign or take the default from class declaration.

```js
constructor({
    coords,
    colorStops = [],
    id,
    ...restOfOptions,
  }: GradientOptions<T>) {

....
Object.assign(this, restOfOptions)

```

why going for the explicit way?

Reply:
I don't have a good reason

Reply:
accept that it keeps junk passed down in options out of the class

Reply:
is exactly what we do with objects on the otherside.
We allow devs to pass id, name, and other things they may want to use.
For me it needs to be clarified, not  blocked

Reply:
I disagree and TS will disagree as well.
This is the exact use case of a subclass.
Stricter js is part of modern js.
It's how I see it.